src/polyfills/fetch.js

```diff
@@ -0,0 +1,247 @@
+
+    /** @type {?ReadableStream} */
+    this.body = null;
+  }
+
+  /**
+   * Create a copy of the response and return it.
+   * @return {!FetchResponse}
+   */
+  clone() {
+    dev().assert(!this.bodyUsed, 'Body already used');
+    return new FetchResponse(this.xhr_);
+  }
+
+  /**
+   * Drains the response and returns the text.
+   * @return {!Promise<string>}
+   * @private
+   */
+  drainText_() {
+    dev().assert(!this.bodyUsed, 'Body already used');
+    this.bodyUsed = true;
+    return Promise.resolve(this.xhr_.responseText);
+  }
+
+  /**
+   * Drains the response and returns a promise that resolves with the response
+   * text.
+   * @return {!Promise<string>}
+   */
+  text() {
+    return this.drainText_();
+  }
+
+  /**
+   * Drains the response and returns the JSON object.
+   * @return {!Promise<!JsonObject>}
+   */
+  json() {
+    return /** @type {!Promise<!JsonObject>} */ (
+      this.drainText_().then(parseJson));
+  }
+
+  /**
+   * Reads the xhr responseXML.
+   * @return {!Promise<!Document>}
+   * @visibleForTesting
+   * @private
+   */
+  document_() {
+    dev().assert(!this.bodyUsed, 'Body already used');
+    this.bodyUsed = true;
+    user().assert(this.xhr_.responseXML,
+        'responseXML should exist. Make sure to return ' +
+        'Content-Type: text/html header.');
+    const doc = /** @type {!Document} */(dev().assert(this.xhr_.responseXML));
+    return Promise.resolve(doc);
+  }
+
+  /**
+   * Drains the response and returns a promise that resolves with the response
+   * ArrayBuffer.
+   * @return {!Promise<!ArrayBuffer>}
+   */
+  arrayBuffer() {
+    return /** @type {!Promise<!ArrayBuffer>} */ (
+      this.drainText_().then(utf8Encode));
+  }
+}
+
+/**
+ * Provides access to the response headers as defined in the Fetch API.
+ * @private Visible for testing.
+ */
+class FetchResponseHeaders {
+  /**
+   * @param {!XMLHttpRequest|!XDomainRequest} xhr
+   */
+  constructor(xhr) {
+    /** @private @const {!XMLHttpRequest|!XDomainRequest} */
+    this.xhr_ = xhr;
+  }
+
+  /**
+   * @param {string} name
+   * @return {string}
+   */
+  get(name) {
+    return this.xhr_.getResponseHeader(name);
+  }
+
+  /**
+   * @param {string} name
+   * @return {boolean}
+   */
+  has(name) {
+    return this.xhr_.getResponseHeader(name) != null;
+  }
+}
+
+/**
+ * @param {string} method
+ * @param {string} url
+ * @return {!XMLHttpRequest|!XDomainRequest}
+ * @private
+ */
+function createXhrRequest(method, url) {
+  let xhr = new XMLHttpRequest();
+  if ('withCredentials' in xhr) {
+    xhr.open(method, url, true);
+  } else if (typeof XDomainRequest != 'undefined') {
+    // IE-specific object.
+    xhr = new XDomainRequest();
+    xhr.open(method, url);
+  } else {
+    throw dev().createExpectedError('CORS is not supported');
+  }
+  return xhr;
+}
+
+/**
+* Adds fetch polyfill if it does not exist.
+* @param {!Window} win
+*/
+export function install(win) {
+  if (!'fetch' in win) {
```

Comment:
Both Travis and LGTM are warning about this line (and 241 and 244), because the condition should be:
```javascript
if (!('fetch' in win))
```
As written, the `!` just applies to `'fetch'`. As `!'fetch'` evaluates to `false`, the current code is the same as writing `false in win`. This means the polyfill is currently _always_ applied, even if the user's browser already supports `fetch`, `Response`, and `Headers`.